### DISCUSSION TITLE
Custom methods on Csla objects and 'Using' relationship

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#2826] [Questions]
I am not sure if I had asked this question in the older forum : )
We are planning to use a 'FeeCalculator' class in our main business object.  We keep a copy of Using Csla books around and on page 36 of Creating business objects book @rockfordlhotka  has explained the 'Using' relationship.
We are discussing a couple of ways to design this

1) We are not sure whether to make FeeCalculator a separate object or make it a business rule which will be called from our main business object? right now we are inclined to code it as a separate object.

2) If we end up with a separate csla object for Fee Calculation, should we have a regular method in our business object or a Csla method?
 By regular method I mean something like SetPersonName() on page 36 (Creating business objects book)

```cs
public void SetPersonName(int personId, string name) 
   {
           var person = PersonEdit.GetPerson(personId); 
           person.FirstName = name; person = person.Save(); 
   }
```

and a Csla method would like this

   ```cs
     public static MethodInfo SetNameMethod = RegisterMethod(c => c.SetName("firstname", "lastname"));
        public void SetName(string firstname, string lastname)
        {
            // Check Authorization, trows UserNotAuthorizedException if not authorized
            CanExecuteMethod(SetNameMethod, true);

            // implementation goes here 

        }
 ```
 where the values "firstname" and "lastname" are dummy values (according to the older forum lol)

Kind Regards

**Comment by rockfordlhotka** [comment]
In current CSLA I think it depends on how the method (because really it sounds like we're talking about a method that calculates a fee) is implemented in your codebase.

1. You can wrap the method in a command object. This is valuable if the logical/physical location of the code might need to be different from the location of the caller. For example, perhaps the fee calculator needs access to the database, so it can only run on the server - so implementing the code in a command object's execute operation method will force it to be on the server to run.
1. You can wrap the method in a rule. This is valuable if the code is location neutral (can run on client or server).
1. You can do both. Implement the calculation in a command object, and invoke that command object from a rule.

I'd probably favor 2 or 3 over 1 these days.
